File: packages/plugin-agent/src/server/memory.ts
/**
 * Agent Memory — retrieves relevant past conversations to inject as context.
 */

export async function getMemoryContext(
  prisma: any,
  userId: string,
  currentMessage: string,
): Promise<string | undefined> {
  // Fetch recent conversations with summaries
  const conversations = await prisma.agentConversation.findMany({
    where: { userId, summary: { not: null } },
    select: { id: true, title: true, summary: true, updatedAt: true },
    orderBy: { updatedAt: 'desc' },
    take: 10,
  });

  if (conversations.length === 0) return undefined;

  // Simple keyword matching for now — can be upgraded to embeddings later
  const queryWords = currentMessage.toLowerCase().split(/\s+/).filter((w: string) => w.length > 3);

  const scored = conversations.map((conv: any) => {
    const text = `${conv.title || ''} ${conv.summary || ''}`.toLowerCase();
    const score = queryWords.reduce((s: number, w: string) => s + (text.includes(w) ? 1 : 0), 0);
    return { ...conv, score };
  });

  const relevant = scored
    .filter((c: any) => c.score > 0)
    .sort((a: any, b: any) => b.score - a.score)
    .slice(0, 3);

  if (relevant.length === 0) return undefined;

  return relevant
    .map((c: any) => `[${c.title || 'Untitled'}]: ${c.summary}`)
    .join('\n');
}

/**
 * Auto-generate a title and summary for a conversation after a few messages.
 */
export async function generateConversationSummary(
  prisma: any,
  conversationId: string,
): Promise<void> {
  const conversation = await prisma.agentConversation.findUnique({
    where: { id: conversationId },
    include: {
      messages: {
        orderBy: { createdAt: 'asc' },
        take: 6,
        select: { role: true, content: true },
      },
    },
  });

  if (!conversation || conversation.messages.length < 2) return;
  if (conversation.title && conversation.summary) return;

  // Simple heuristic: use first user message as title, concatenate for summary
  const firstUserMsg = conversation.messages.find((m: any) => m.role === 'user');
  const title = firstUserMsg
    ? (firstUserMsg.content as string).slice(0, 100)
    : 'Agent conversation';

  const summary = conversation.messages
    .filter((m: any) => m.role === 'user' || m.role === 'assistant')
    .map((m: any) => `${m.role}: ${(m.content as string).slice(0, 200)}`)
    .join(' | ')
    .slice(0, 500);

  await prisma.agentConversation.update({
    where: { id: conversationId },
    data: { title, summary },
  });
}

File: apps/web/apps/agent/memory.py
"""
Agent memory — retrieves relevant past conversations for context injection.

Port of packages/plugin-agent/src/server/memory.ts
"""

from __future__ import annotations

from .models import AgentConversation, AgentMessage


def get_memory_context(user_id: str, current_message: str) -> str | None:
    """Return a short text block with relevant past conversation summaries."""

    conversations = (
        AgentConversation.objects.filter(user_id=user_id)
        .exclude(summary__isnull=True)
        .exclude(summary="")
        .order_by("-updated_at")[:10]
    )

    if not conversations:
        return None

    # Simple keyword matching (upgradeable to embeddings later)
    query_words = [w.lower() for w in current_message.split() if len(w) > 3]
    if not query_words:
        return None

    scored = []
    for conv in conversations:
        text = f"{conv.title or ''} {conv.summary or ''}".lower()
        score = sum(1 for w in query_words if w in text)
        if score > 0:
            scored.append((score, conv))

    scored.sort(key=lambda x: x[0], reverse=True)
    relevant = scored[:3]

    if not relevant:
        return None

    return "\n".join(
        f"[{conv.title or 'Untitled'}]: {conv.summary}"
        for _, conv in relevant
    )


def generate_conversation_summary(conversation_id: str) -> None:
    """Auto-generate a title and summary for a conversation after a few messages."""

    try:
        conv = AgentConversation.objects.get(id=conversation_id)
    except AgentConversation.DoesNotExist:
        return

    msgs = list(
        conv.messages.order_by("created_at")[:6].values_list("role", "content")
    )
    if len(msgs) < 2:
        return
    if conv.title and conv.summary:
        return

    # Simple heuristic: first user message as title
    first_user = next((content for role, content in msgs if role == "user"), None)
    title = (first_user[:100] if first_user else "Agent conversation")

    summary = " | ".join(
        f"{role}: {content[:200]}"
        for role, content in msgs
        if role in ("user", "assistant")
    )[:500]

    AgentConversation.objects.filter(id=conversation_id).update(title=title, summary=summary)
